File: problems/2025/08-30/the-celebrity-problem/solution.js
function celebrity(mat) {
  const n = mat.length;
  if (n === 0) return -1;

  let i = 0, j = n - 1;
  while (i < j) {
    if (mat[i][j] === 1) {
      i++;
    } else {
      j--;
    }
  }

  const cand = i;
  for (let k = 0; k < n; k++) {
    if (k === cand) continue;
    if (mat[cand][k] === 1 || mat[k][cand] === 0) return -1;
  }
  return cand;
}

// quick checks
console.log(celebrity([[1,1,0],[0,1,0],[0,1,1]])); // 1
console.log(celebrity([[1,1],[1,1]]));             // -1
console.log(celebrity([[1]]));                     // 0

File: problems/2025/08-30/the-celebrity-problem/solution.py
from typing import List


def celebrity(mat: List[List[int]]) -> int:
    n = len(mat)
    if n == 0:
        return -1

    i, j = 0, n - 1
    while i < j:
        if mat[i][j] == 1:
            i += 1
        else:
            j -= 1

    cand = i

    for k in range(n):
        if k == cand:
            continue
        if mat[cand][k] == 1 or mat[k][cand] == 0:
            return -1

    return cand


if __name__ == "__main__":
    print(celebrity([[1,1,0],[0,1,0],[0,1,1]]))  # 1
    print(celebrity([[1,1],[1,1]]))              # -1
    print(celebrity([[1]]))                      # 0


